interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,337 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+)
+
+const mountControlSummary = `allows creating transient and persistent mounts`
+
+const mountControlBaseDeclarationPlugs = `
+  mount-control:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const mountControlBaseDeclarationSlots = `
+  mount-control:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-connection: true
+`
+
+const mountAttrNotADictionary = `mount-control "mount" attribute is not a dictionary`
+const mountControlConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access.
+mount
+umount
+umount2
+`
+
+var allowedMountOptions = []string{
+	"async",
+	"atime",
+	"bind",
+	"diratime",
+	"dirsync",
+	"iversion",
+	"lazytime",
+	"nofail",
+	"noiversion",
+	"nomand",
+	"noatime",
+	"nodev",
+	"nodiratime",
+	"noexec",
+	"nolazytime",
+	"norelatime",
+	"nosuid",
+	"nostrictatime",
+	"nouser",
+	"relatime",
+	"strictatime",
+	"sync",
+	"ro",
+	"rw",
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// TODO: review the regexp
+// We prohibit expressions containing a comma, as that is used by AppArmor to
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
+
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
```

Comment:
as mentioned below, this is not enough, we should be disallowing certain kinds of filesystems from being mounted